IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m

```diff
@@ -1175,24 +1175,33 @@ - (BOOL)saveAccount:(MSIDAccount *)account
         }
         else
         {
-            [noReturnAccountsSet addObject:account];
+            [noReturnAccountsSet addObject:accountCacheItem.username];
         }
     }
     
-    if (noReturnAccounts) *noReturnAccounts = noReturnAccountsSet;
+    if (noReturnAccountUPNs) *noReturnAccountUPNs = noReturnAccountsSet;
     
     return returnAccountsSet;
 }
 
 - (NSArray<MSIDAccount *> *)filterSignedOutAccountsFromOtherAccessor:(NSArray<MSIDAccount *> *)accounts
-                                                    noReturnAccounts:(NSSet<MSIDAccount *> *)noReturnAccountsSet
+                                                accountMetadataCache:(MSIDAccountMetadataCacheAccessor *)accountMetadataCache
+                                                            clientId:(NSString *)clientId
+                                                 noReturnAccountUPNs:(NSSet<NSString *> *)noReturnAccountUPNSet
 {
     NSMutableArray *returnAccounts = [NSMutableArray new];
     for (MSIDAccount *account in accounts)
     {
-        // All signed out accounts should be appearing in noReturnAccountSet,
-        // so no need to check account metadata again here (assuming we never delete account objects).
-        if ([noReturnAccountsSet containsObject:account]) continue;
+        // We rely on UPN check only if home account id is not available
+        if ([NSString msidIsStringNilOrBlank:account.accountIdentifier.homeAccountId])
+        {
+            if ([noReturnAccountUPNSet containsObject:account.username]) continue;
+        }
+        else
+        {
+            MSIDAccountMetadataState signInState = [accountMetadataCache signInStateForHomeAccountId:account.accountIdentifier.homeAccountId clientId:clientId context:nil error:nil];
```

Comment:
nit: this call is technically done twice for each account (here and above). Would it be possible to optimize?

Reply:
The same call at line 1133 is for default accounts and here is for legacy accounts. To optimize, I can replace line 1200 with
`else if ([returnAccounts containObject:account]) {continue;} else`
What do you think? :)